go-controller/pkg/util/go_ovn.go

```diff
@@ -0,0 +1,776 @@
+	case config.OvnDBSchemeTCP:
+		ovnSBClient.client, err = initGoOvnTcpClient(config.OvnSouth.GetURL(), goovn.DBSB)
+	case config.OvnDBSchemeUnix:
+		ovnSBClient.client, err = initGoOvnUnixClient(config.OvnSouth.GetURL(), goovn.DBSB)
+	default:
+		klog.Errorf("Invalid db scheme: %s when initializing the OVN SB Client",
+			config.OvnSouth.Scheme)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Couldn't initialize SBDB client: %s", err)
+	}
+
+	klog.Infof("Created OVN SB client with Scheme: %s", config.OvnSouth.Scheme)
+	return nil
+}
+
+func initGoOvnSslClient(certFile, privKeyFile, caCertFile, address, db string) (goovn.Client, error) {
+	cert, err := tls.LoadX509KeyPair(certFile, privKeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("Error generating x509 certs for ovndbapi: %s", err)
+	}
+	caCert, err := ioutil.ReadFile(caCertFile)
+	if err != nil {
+		return nil, fmt.Errorf("Error generating ca certs for ovndbapi: %s", err)
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      caCertPool,
+	}
+	tlsConfig.BuildNameToCertificate()
+	ovndbclient, err := goovn.NewClient(&goovn.Config{
+		Db:           db,
+		Addr:         address,
+		TLSConfig:    tlsConfig,
+		DisconnectCB: disconnectCb,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Error creating SSL OVNDBClient for database %s at address %s: %s", db, address, err)
+	}
+	klog.Infof("Created OVNDB SSL client for db: %s", db)
+	return ovndbclient, nil
+}
+
+func initGoOvnTcpClient(address, db string) (goovn.Client, error) {
+	ovndbclient, err := goovn.NewClient(&goovn.Config{
+		Db:           db,
+		Addr:         address,
+		DisconnectCB: disconnectCb,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Error creating TCP OVNDBClient for address %s: %s", address, err)
+	}
+	klog.Infof("Created OVNDB TCP client for db: %s", db)
+	return ovndbclient, nil
+}
+
+func initGoOvnUnixClient(address, db string) (goovn.Client, error) {
+	ovndbclient, err := goovn.NewClient(&goovn.Config{
+		Db:           db,
+		Addr:         address,
+		DisconnectCB: disconnectCb,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Error creating UNIX OVNDBClient for address %s: %s", address, err)
+	}
+	klog.Infof("Created OVNDB UNIX client for db: %s", db)
+	return ovndbclient, nil
+}
+
+func disconnectCb() {
+	klog.Infof("Received disconnect callback for the clients")
+	defer reconnect()
+}
+
+func reconnect() {
+	reconnectTicker := time.NewTicker(500 * time.Millisecond)
+	go func() {
+		for range reconnectTicker.C {
+			klog.Infof("Reconnecting OVN client")
+			if err := InitOVNDBClients(); err != nil {
+				continue
+			}
+			return
+		}
+
+	}()
+}
+
+// Client Interface Methods
+
+// Get logical switch by name
+func (ovnclient *OVNClient) LSGet(ls string) ([]*goovn.LogicalSwitch, error) {
+	if ovnclient.client == nil {
+		return nil, syscall.ENOTCONN
+	}
+	return ovnclient.client.LSGet(ls)
```

Comment:
Why is it possible to have an `*OVNClient` without a `.client`? Seems like that should be handled at a higher level.

(Like, rather than having a global `*OVNClient`, we should initialize one at startup and then pass it to the node/controller object, etc)

Reply:
So whether we need to pass the OVNClient around or keep it global is slightly orthogonal to why I have a `.client` shimmed by `OVNClient`. The primary reason is for connection health-check. The underneath RPC2 client at scale was getting disconnected too easily when ovsdb stopped responding (at high load). There is a reconnect mechanism in place, but we still aren't guaranteed when the commands get invoked, if the client instance is a valid and connected one. So the option is to add some API for health-check everywhere in the code before making calls or set the `.client` to `nil` on disconnect and add a shim layer around the code (which is what it does right now) for error-checking.

Reply:
so then should it be called ovnclient.connection ?

Reply:
semantically it's not the `connection` it's the actual `goovn.Client` interface construct.

Reply:
Wait, so where is the client set to `nil` then? And I still don't understand the point - if the connection is disconnected, you'll get an error one way or another...

Reply:
Earlier, the `reconnect` semantic didn't exist in the `goovn.Client` implementation. In the latest changes, Han added those and I made the change to use the reconnect facility. However, we still need the `OVNClient` because the `goovn.Client` doesn't adhere to the `OVNInterface`. In particular commands need to return `OVNCommandInterface` for mocks to work. So at this point the need to have OVNClient wrapper interface is purely for unit-testing.

Reply:
that's fine, but I still think you can remove the `nil` check.

Reply:
yeah sure.